### ISSUE TITLE
Directive for force-compiling code branches for different bint-parameters?

### ISSUE THREAD

**Issue by h-vetinari** [cython/cython#2660]
Not sure if this is feasible, but I had the following situation while working on pandas hashtable code recently (simplified a bit from pandas-dev/pandas#22986 for simplicity):

Two methods had substantial code overlap, with the difference being easily absorbed into a `bint` parameter - in this case, whether the `unique` method returns an inverse or not. It would be very desirable from a clarity/readability/maintenance perspective to unify the code for those two methods.

I tried unifying these functions, but there was a perf-hit for the simpler case (no inverse) of about 10%, which is too much for just always calculating it by default for such a central method.

I then tried (very naively) to get cython to compile the different code branches as follows:

```
    def _unique(self, ndarray[object] values, bint return_inverse=False):
        [...]
        with nogil:
            for i in range(n):
                [actual computation]
                if return_inverse:
                    [calculate inverse]

    def unique(self, ndarray[object] values, bint return_inverse=False):
        # define separate functions with/without inverse to force compilation
        # of the different code paths for boolean "return_inverse"
        if return_inverse:
            return self._unique_with_inverse(values)
        return self._unique_no_inverse(values)

    def _unique_no_inverse(self, const {{dtype}}_t[:] values):
        return self._unique(values, return_inverse=False)

    def _unique_with_inverse(self, const {{dtype}}_t[:] values):
        return self._unique(values, return_inverse=True)
```
but the perf penalty remained.

Summing up, it would be awesome to be able to do something like (the example below borrows from `@pytest.mark.parametrize`, but the exact syntax is not relevant to me):
```
    @cython.force_compile('return_inverse', [False])
    def unique(self, ndarray[object] values, bint return_inverse=False):
        [...]
        with nogil:
            for i in range(n):
                [compute]
                if return_inverse:
                    [calculate inverse]
```
with the idea being to *both* unify the code *and* maintain perf in the case where the `return_inverse`-branch doesn't need to be touched at all.

**Comment by robertwb** [comment]
If I understand correctly, you're asking for a feature to generate multiple versions of a codepath (like templates) for a list of specified values, as well as the generic one. This sounds like a complicated addition to the language best taken up on the mailing list. 

However, for your use, I don't think you'd need it. First, given your use of {{dtype}} I assume you're already using templates to auto-generate code at a higher level, so it shouldn't be too bad to leverage this templating functionality to explicitly generate to copies of this function parameterized on {{return_inverse}}. If you need to.

I would be surprised if checking a bint, even millions of times, in a loop really has noticeable performance given how fast processors are these days (especially with branch prediction). Given the snippet above, I suspect the speed loss is due to calling multiple def functions, passing return_inverse as a keyword argument, and typing values as a memory view (requiring buffer unpacking) in the wrapper methods that just pass it on. Of course I'm just going on eyeballing the snippets in this ticket, without running any profiles myself, so that hypothesis would need to be validated against the actual code.

**Comment by h-vetinari** [comment]
@robertwb Thanks for the response!

You did understand correctly, and you're right - it is pretty much exactly like templating. Not sure how easy it would be to auto-template on a variable - I guess a necessary restriction would be, that the value needs to be `const`.

In general, the ``{{dtype}}`` (and other) templating is not the most ideal solution - for the former, there's fused types at least, which allows to write code much more natively.

So for the latter, it would be great to be able to avoid templating as well; e.g. do something like
```
@cython.force_compile('parameter_c', [val1, val2])
def some_function(parameter_a, parameter_b, parameter_c)
    [do stuff]
```
and have it *internally* compile to the equivalent of
```
def some_function(parameter_a, parameter_b, parameter_c)
    if parameter_c == val1:
        return some_function_fc_p2_v0(parameter_a, parameter_b)
    elif parameter_c == val2:
        return some_function_fc_p2_v1(parameter_a, parameter_b)
    # general case
    [do stuff]
```
where `some_function_fc_p2_v0` is **f**orce **c**ompiled with **p**arameter 2 set to **v**alue 0, etc.

**Comment by h-vetinari** [comment]
Regarding your comment about my concrete case:
> I would be surprised if checking a bint, even millions of times, in a loop really has noticeable performance given how fast processors are these days (especially with branch prediction). [...] Of course I'm just going on eyeballing the snippets in this ticket, without running any profiles myself, so that hypothesis would need to be validated against the actual code.

here are the relevant code path from pandas-dev/pandas#22986, commit https://github.com/pandas-dev/pandas/pull/22986/commits/9918d52b96f722495fb1a72135e991d725ba3cda for the attempted unification and https://github.com/pandas-dev/pandas/pull/22986/commits/dbe4e0ed81e826ca631c7b70c0117b5026a5f079 for reverting back to separate code path.

[Note: I wrote this response as extracts-first-response-later, but am now switching that around, to maybe save you some time]

From your list:
>  Given the snippet above, I suspect the speed loss is due to calling multiple def functions, passing return_inverse as a keyword argument, and typing values as a memory view (requiring buffer unpacking) in the wrapper methods that just pass it on

So I don't think that it's multiple `def`, or any memory view differences [see below] - but do I understand you correctly that simply passing the arguments as keywords rather than positionally could be responsible for the perf loss? That would have been a complete blind spot for me - guess I still have many cython intricacies left to learn...

[\Note]

Here's an extract of the relevant code for the unification - which ended up ~10% slower. The only relevant parts here are the `bint` for `return_inverse` and `ignore_na` (`pandas.factorize` removes NaNs from the uniques, while `pandas.unique` doesn't). The latter was always False for the performance tests in question, so just another case of reading a `bint`. 

```
    def unique(self, const {{dtype}}_t[:] values, bint return_inverse=False):
        return self._unique(values, uniques={{name}}Vector(), ignore_na=False,
                            return_inverse=return_inverse)

    @cython.boundscheck(False)
    def _unique(self, const {{dtype}}_t[:] values, {{name}}Vector uniques,
                bint ignore_na=False, bint return_inverse=False,
                Py_ssize_t count_prior=0, Py_ssize_t na_sentinel=-1,
                object na_value=None):
        cdef:
            Py_ssize_t i, idx, count = count_prior, n = len(values)
            int64_t[:] labels
            int ret = 0
            {{dtype}}_t val, na_value2
            khiter_t k
            {{name}}VectorData *ud
            bint use_na_value

        if return_inverse:
            labels = np.empty(n, dtype=np.int64)
        ud = uniques.data
        use_na_value = na_value is not None

         if use_na_value:
             # We need this na_value2 because we want to allow users
             # to *optionally* specify an NA sentinel *of the correct* type.
             # We use None, to make it optional, which requires `object` type
             # for the parameter. To please the compiler, we use na_value2,
             # which is only used if it's *specified*.
             na_value2 = <{{dtype}}_t>na_value
         else:
             na_value2 = {{default_na_value}}
 
         with nogil:
            for i in range(n):
                val = values[i]
                if ignore_na and (val != val
                                  or (use_na_value and val == na_value2)):
                    labels[i] = na_sentinel
                    continue

                k = kh_get_{{dtype}}(self.table, val)
                if return_inverse and k != self.table.n_buckets:
                    # k falls into a previous bucket
                    idx = self.table.vals[k]
                    labels[i] = idx
                elif k == self.table.n_buckets:
                    # k hasn't been seen yet
                    k = kh_put_{{dtype}}(self.table, val, &ret)

                    if needs_resize(ud):
                        with gil:
                            if uniques.external_view_exists:
                                 raise ValueError("external reference to "
                                                  "uniques held, but "
                                                 "Vector.resize() needed")
                            uniques.resize()
                    append_data_{{dtype}}(ud, val)
                    if return_inverse:
                        self.table.vals[k] = count
                        labels[i] = count
                    count += 1

        if return_inverse:
            return uniques.to_array(), np.asarray(labels)
        return uniques.to_array()
```

whereas the code that was faster did just:
```
    def unique(self, const {{dtype}}_t[:] values, bint return_inverse=False):
        if return_inverse:
            return self._unique(values, uniques={{name}}Vector(), ignore_na=False,
                                return_inverse=True)
        return self._unique_no_inverse(values)

    @cython.boundscheck(False)
    def _unique_no_inverse(self, const {{dtype}}_t[:] values):
        # define separate functions without inverse for performance
        cdef:
           Py_ssize_t i, n = len(values)
           int ret = 0
           {{dtype}}_t val
           khiter_t k
           {{name}}Vector uniques = {{name}}Vector()
           {{name}}VectorData *ud
        ud = uniques.data
        with nogil:
            for i in range(n):
                val = values[i]
                k = kh_get_{{dtype}}(self.table, val)
                if k == self.table.n_buckets:
                    kh_put_{{dtype}}(self.table, val, &ret)
                    if needs_resize(ud):
                        with gil:
                            uniques.resize()
                    append_data_{{dtype}}(ud, val)
        return uniques.to_array()
```

**Comment by robertwb** [comment]
Passing arguments by keyword is slower in Python, which we inherit. This may, of course, not be the issue if the called code is where the bulk of the time is spent. 

If you could whittle this down to a self-contained example that still exhibits the issue, I could try and see where exactly the slowdown is coming from.

**Comment by h-vetinari** [comment]
@robertwb 
I've tried removing all the kwargs in pandas-dev/pandas#23400, but to no effect. The templating works, but is ugly/overkill, IMO.

Here's an attempt at a pared-down minimal example, directly from that PR. I'm showing the diff between the original version that was directly compiled, and the one that's 10% slower. From what I can tell, the only difference is some type declarations (but not allocations), and checking two `bint` throughout the loop (I've manually removed some lines due to handling `na_value` that get never hit if `ignore_na=False`).

```diff
     @cython.boundscheck(False)
     @cython.wraparound(False)
-    def unique(self, const {{dtype}}_t[:] values):
+    def _unique(self, const {{dtype}}_t[:] values, {{name}}Vector uniques,
+                bint ignore_na, bint return_uniques,
+                bint return_inverse, Py_ssize_t count_prior,
+                Py_ssize_t na_sentinel, object na_value):
         cdef:
-            Py_ssize_t i, n = len(values)
+            Py_ssize_t i, n = len(values), idx, count = count_prior
+            {{dtype}}_t[:] labels
             int ret = 0
             {{dtype}}_t val
             khiter_t k
-            {{name}}Vector uniques = {{name}}Vector()
             {{name}}VectorData *ud

+        if return_inverse:
+            labels = np.empty(n, dtype=np.int64)
         ud = uniques.data

         with nogil:
             for i in range(n):
                 val = values[i]
+
+                if ignore_na and (val != val or val == na_value):
+                    labels[i] = na_sentinel
+                    continue
+
                 k = kh_get_{{dtype}}(self.table, val)
                 if k == self.table.n_buckets:
-                    kh_put_{{dtype}}(self.table, val, &ret)
+                    # k hasn't been seen yet
+                    k = kh_put_{{dtype}}(self.table, val, &ret)
+
                     if needs_resize(ud):
                         with gil:
                             uniques.resize()
                     append_data_{{dtype}}(ud, val)
-        return uniques.to_array()
+                    if return_inverse:
+                        self.table.vals[k] = count
+                        labels[i] = count
+                        count += 1
+                elif return_inverse:
+                    # k falls into a previous bucket
+                    # only relevant in case we need to construct the inverse
+                    idx = self.table.vals[k]
+                    labels[i] = idx
+
+        if return_uniques and return_inverse:
+            return uniques.to_array(), np.asarray(labels)
+        elif return_uniques:
+            return uniques.to_array()
+        elif return_inverse:
+            return np.asarray(labels)
```

There's one more call-layer, but that should not be an issue, I'm guessing:
```
    def unique(self, const {{dtype}}_t[:] values, bint return_inverse):
        uniques = {{name}}Vector()
        return self._unique(values, uniques,
                            False,  # ignore_na
                            True,   # return_uniques
                            return_inverse,
                            # the rest are of the parameters are not relevant,
                            # but we don't use kwargs to avoid cython perf hit
                            0,      # count_prior
                            -1,     # na_sentinel
                            None)   # na_value
```

**Comment by robertwb** [comment]
Is there an example that reproduces the issue that doesn't require
templating? (Also, I don't know what the {{name}}Vector types are, are
these critical to the issue as well?) I don't care if it computes the right
thing so long as it reproduces the performance issues.

**Comment by jakirkham** [comment]
FWIW there is actually a neat trick that one can do in Cython using fused types, which is similar to [C++'s SFINAE]( https://en.cppreference.com/w/cpp/language/sfinae ) like [what has been used with `enable_if`]( https://www.boost.org/doc/libs/1_73_0/libs/core/doc/html/core/enable_if.html#core.enable_if.using_enable_if ). Here's a short example of how this works below.

```cython
cdef struct bint_false_type:
    char empty                               # need something as empty `struct`s are not valid

cdef struct bint_true_type:
    char empty                               # ditto

cdef fused param_bint_type:                  # fuse `True` and `False` types to allow selection between them
     bint_false_type
     bint_true_type


cdef inline bint cfunc(param_bint_type* p):  # include unused argument of fused type for selection
    if param_bint_type is bint_true_type:    # compile time dispatching on type
        return True
    else:
        return False

def func(bint b):
    if b:
        return cfunc(<bint_true_type*>NULL)  # cast `NULL` pointer to needed type for selection
    else:
        return cfunc(<bint_false_type*>NULL)
```

**Comment by h-vetinari** [comment]
Thanks for the tip @jakirkham, maybe I'll have some time to improve that old pandas PR. :)
